apps/l10nstats/management/commands/progress.py

```diff
@@ -108,18 +120,38 @@ def handle(self, *args, **options):
         ProgressPosition.objects.all().delete()
         ProgressPosition.objects.bulk_create(backobjs)
 
-    def rescale(self, vals):
-        _min = min(c for _, c in vals)
-        _max = max(c for _, c in vals)
-        if ((_min <= 25 and _max >= 75) or
+    def rescale(self, vals, span=.75):
+        # return a scaling function for change values
+        # Make graph at least "span" % high
+        # Ensure that upper and lower space have the same
+        # proportions as the original graph
+        # The scaling function return a value 0 <= v <= 1
+        _min = min(c for _, c, __ in vals)
+        _max = max(c for _, c, __ in vals)
+        total = max(t for _, __, t in vals)
+        if ((_max - _min) >= total * span or
             (_min >= _max)):
             # no resize needed or useful
-            return vals
-        bottom = _min if _min <= 25 else 25
-        top = _max if _max >= 75 else 75
-        ratio = float(top - bottom) / (_max - _min)
-        scale = lambda v: (v - _min) * ratio + bottom
-        return [(t, scale(c)) for t, c in vals]
+            return lambda v: 1.0 * v / total
+        ratio = span / (_max - _min)
+        offset = (1.0 - span) * _min / (total - _max + _min)
+        return lambda v: (v - _min) * ratio + offset
+
+    def initialCoverage(self, tree2locs, startdate):
+        rv = {}
+        for tree, locales in tree2locs.iteritems():
+            locs = (Locale.objects
+                .filter(code__in=locales)
+                .filter(run__srctime__lt=startdate, run__tree__code=tree)
```

Comment:
I'm not sure if this will work with the reverse relationship you're using here, but you may want to use `prefetch_related` or `select_related` with `run` to get the run relations as a separate but individual query or as an inner query. That should make this type of iterative query faster.